hudi-io/src/main/java/org/apache/hudi/storage/StorageConfiguration.java

```diff
@@ -0,0 +1,132 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.hudi.storage;
+
+import org.apache.hudi.common.util.Option;
+import org.apache.hudi.common.util.StringUtils;
+
+import java.io.IOException;
+import java.io.ObjectInputStream;
+import java.io.ObjectOutputStream;
+import java.io.Serializable;
+
+/**
+ * Interface providing the storage configuration in type {@link T}.
+ *
+ * @param <T> type of storage configuration to provide.
+ */
+public abstract class StorageConfiguration<T> implements Serializable {
+  /**
+   * @return the storage configuration.
+   */
+  public abstract T get();
+
+  /**
+   * @return a new copy of the storage configuration.
+   */
+  public abstract T newCopy();
+
+  /**
+   * Serializes the storage configuration.
+   * DO NOT change the signature, as required by {@link Serializable}.
+   *
+   * @param out stream to write.
+   * @throws IOException on I/O error.
+   */
+  public abstract void writeObject(ObjectOutputStream out) throws IOException;
```

Comment:
Does this need to be `ObjectOutputStream` due to `Serializable`, right? We should have ability to control the binary serialization of this object , lets make sure of that?

Reply:
Yes, the method signature has to be like this for `Serializable` and the implementation has to provide a custom one.